Source/EasyNetQ/Consumer/ConsumerDispatcher.cs

```diff
@@ -8,53 +8,73 @@ namespace EasyNetQ.Consumer
     public class ConsumerDispatcher : IConsumerDispatcher
     {
         private readonly ILog logger = LogProvider.For<ConsumerDispatcher>();
-        private readonly BlockingCollection<Action> queue;
-        private bool disposed;
+        private readonly AutoResetEvent autoResetEvent = new AutoResetEvent(false);
+        private readonly ConcurrentQueue<Action> durableActions = new ConcurrentQueue<Action>();
+        private readonly ConcurrentQueue<Action> transientActions = new ConcurrentQueue<Action>();
 
         public ConsumerDispatcher(ConnectionConfiguration configuration)
         {
             Preconditions.CheckNotNull(configuration, "configuration");
 
-            queue = new BlockingCollection<Action>();
-
             var thread = new Thread(_ =>
             {
-                while (!disposed && queue.TryTake(out var action, -1))
-                {
+                while (!IsDone())
                     try
                     {
-                        action();
+                        if (durableActions.TryDequeue(out var action) || transientActions.TryDequeue(out action))
+                            action();
+                        else
+                            autoResetEvent.WaitOne();
```

Comment:
```suggestion
                      if (!IsDisposed) <-- fast check to ensure not disposed yet
                            autoResetEvent.WaitOne(10000);  <-- timeout - guard from deadlock
```

Race conditions with `Dispose()`. Will block forever If called after `Dispose()`.